components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketInboundHandler.java

```diff
@@ -585,4 +588,28 @@ private void removeTokenFromQuery(Map<String, List<String>> parameters) {
         // remove trailing '?' or '&' from the built string
         uri = queryBuilder.substring(0, queryBuilder.length() - 1);
     }
+    private SignedJWTInfo getSignedJwtInfo(String accessToken) throws ParseException {
+
+        String signature = accessToken.split("\\.")[2];
+        SignedJWTInfo signedJWTInfo = new SignedJWTInfo();
+        Cache gatewaySignedJWTParseCache = CacheProvider.getGatewaySignedJWTParseCache();
+        if (gatewaySignedJWTParseCache != null) {
+            Object cachedEntry = gatewaySignedJWTParseCache.get(signature);
+            if (cachedEntry != null) {
+                signedJWTInfo = (SignedJWTInfo) cachedEntry;
+            } else {
+                SignedJWT signedJWT = SignedJWT.parse(accessToken);
+                signedJWTInfo.setSignedJWT(signedJWT);
```

Comment:
This part should go into SignedJWTInfo. So SignedJWTInfo has a constructor which accepts a string constructor. Inside it the the signedJWT is created. Then while calling to SignedJWTInfo.getClaimSet method, the getClaimSet method of the signedJWT is called and resulting claims are cached at the SignedJWTInfo layer. getClaimSet is not called prematurely but only when it's needed. And the ClaimSet is saved for subsequent calls.